src/meta/src/rpc/server.rs

```diff
@@ -171,308 +158,141 @@ pub async fn rpc_serve_with_store<S: MetaStore>(
+        // FIXME: Add service discovery for follower
+        // https://github.com/risingwavelabs/risingwave/issues/6755
+        let mut svc_shutdown_rx_clone = svc_shutdown_rx.clone();
+        let (follower_shutdown_tx, follower_shutdown_rx) = OneChannel::<()>();
+        let (follower_finished_tx, follower_finished_rx) = OneChannel::<()>();
+        if !is_leader {
             tokio::spawn(async move {
+                let _ = tracing::span!(tracing::Level::INFO, "follower services").enter();
+                tracing::info!("Starting follower services");
+                let health_srv = HealthServiceImpl::new();
                 tonic::transport::Server::builder()
                     .layer(MetricsMiddlewareLayer::new(Arc::new(MetaMetrics::new())))
                     .add_service(HealthServer::new(health_srv))
-                    .serve(address_info.listen_addr)
+                    .serve_with_shutdown(address_info.listen_addr, async move {
+                        tokio::select! {
+                            _ = tokio::signal::ctrl_c() => {},
+                            // shutdown service if all services should be shut down
+                            res = svc_shutdown_rx_clone.changed() =>  {
+                                match res {
+                                    Ok(_) => tracing::info!("Shutting down services"),
+                                    Err(_) => tracing::error!("Service shutdown sender dropped")
+                                }
+                            },
+                            // shutdown service if follower becomes leader
+                            res = follower_shutdown_rx =>  {
+                                match res {
+                                    Ok(_) => tracing::info!("Shutting down follower services"),
+                                    Err(_) => tracing::error!("Follower service shutdown sender dropped")
+                                }
+                            },
+                        }
+                    })
                     .await
                     .unwrap();
+                match follower_finished_tx.send(()) {
+                    Ok(_) => tracing::info!("Shutting down follower services done"),
+                    Err(_) => {
+                        tracing::error!("Follower service shutdown done sender receiver dropped")
+                    }
+                }
             });
-        });
-
-        let shutdown_election = async move {
-            if election_shutdown.send(()).is_err() {
-                tracing::warn!("election service already shut down");
-            } else if let Err(err) = election_handle.await {
-                tracing::warn!("Failed to join shutdown: {:?}", err);
-            }
-        };
+        }
 
-        let (svc_shutdown_tx, mut svc_shutdown_rx) = tokio::sync::oneshot::channel::<()>();
+        // wait until this node becomes a leader
+        while !services_leader_rx.borrow().clone().1 {
```

Comment:
I think we need some wrapper functions for leader info and info watch logic to make the mainline logic more concise and clear.

Reply:
I added below function. I hope this simplifies the code a little bit. 

```rust
fn node_is_leader(leader_rx: &WatchReceiver<(MetaLeaderInfo, bool)>) -> bool {
    leader_rx.borrow().clone().1
}
```